### PR TITLE
[Feature] Base version of inverted index

### LINKED ISSUE #34019: Support fulltext search with inverted index
## Feature request
**Motivation**
Inverted index is a commonly used data structure in information retrieval systems, which maps the "word-document" relationship to the "document-word" relationship. In an inverted index, each unique word has an associated index pointing to a list of documents containing that word. This indexing method makes document search based on words highly efficient.

In OLAP engines, inverted indexes and information retrieval play a crucial role. OLAP engines are typically used for handling complex analytical queries, which often involve large amounts of data and require aggregation and filtering across multiple dimensions.

Inverted indexes can greatly improve the performance of these queries. For example, if you want to find all data rows containing a specific word, you can directly look up the inverted index instead of scanning the entire dataset. This can significantly optimize the query scan volume, especially when retrieving a small amount of data from massive datasets, resulting in greatly improved query performance.

Currently, several OLAP engines, such as `Doris` and `ClickHouse`, have their own implementation of inverted indexes.

**Solution**

We want to abstract an implementation framework for inverted indexes, and CLucene is the first specific implementation we are trying.
flowchart (1)

When we use inverted index, we can use grammar below to manage, write and query index.
```
// create table with inverted index
CREATE TABLE table_name
(
  columns_difinition,
  INDEX idx_name1(column_name1) USING GIN [PROPERTIES("analyzer" = "english|chinese|none", ...)] [COMMENT 'your comment']
  INDEX idx_name2(column_name2) USING GIN [PROPERTIES("analyzer" = "english|chinese|none", ...)] [COMMENT 'your comment']
)

// create index for table
-- grammar 1
CREATE INDEX idx_name ON table_name(column_name) USING GIN [PROPERTIES("analyzer" = "english|chinese|none", ...)] [COMMENT 'your comment'];
-- grammar 2
ALTER TABLE table_name ADD INDEX idx_name(column_name) USING GIN [PROPERTIES("analyzer" = "english|chinese|none", ...)] [COMMENT 'your comment'];
-- grammar 3 
ALTER TABLE table_name MODIFY INDEX idx_name(column_name)  [PROPERTIES] [COMMENT 'your comment']

// drop index
-- grammar 1
DROP INDEX idx_name ON table_name;
-- grammar 2
ALTER TABLE table_name DROP INDEX idx_name;

// query match inverted index with basic binary comparison operators: =,!=,>,>=,<,<=
select * from table_name where column='xxx'; // match

// furthermore, we can also use the following functions to have complete full-text search ability.
// terms query
select * from table where match_terms(f, ['n1', 'n2']);

// phrase query
select * from table where match_phrase(f1, "n1", "n2", 2);

// wildcard query
select * from table where match_wildcard(f1, "*15*");

// fuzzy query
select * from table where match_fuzzy(f1, "thinkbad", 1);

// range query
select * from table where match_range(f1, 1, 10, false, true);

// boost for query
select * from table where boost(terms_query(name, ['n1']), 10) and [other conditions];
```

### PR DESCRIPTION
**Pull request by dujijun007**
Fixes #34019

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [x] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by imay** [comment]
@dujijun007 Thank you so much for contributing this significant feature for starrocks.
This change is so big that it will be more readable if you can split this into multiple smaller one.
For example, could you submit the thirdparty first?  After this PR is merged, dev-image will be updated with a new thirdpary so other developers won't suffer compiling issues.

**Comment by dujijun007** [comment]
> @dujijun007 Thank you so much for contributing this significant feature for starrocks. This change is so big that it will be more readable if you can split this into multiple smaller one. For example, could you submit the thirdparty first? After this PR is merged, dev-image will be updated with a new thirdpary so other developers won't suffer compiling issues.

**Comment by dujijun007** [comment]
> @dujijun007 Thank you so much for contributing this significant feature for starrocks. This change is so big that it will be more readable if you can split this into multiple smaller one. For example, could you submit the thirdparty first? After this PR is merged, dev-image will be updated with a new thirdpary so other developers won't suffer compiling issues.

@imay I separate the first PR for dependency here #34134

File: be/src/storage/rowset/segment_iterator.cpp [hunk 1]
```diff
@@ -1834,6 +1844,57 @@ Status SegmentIterator::_apply_del_vector() {
...
+        for (const ColumnPredicate* pred : pred_list) {
+            Status res = pred->seek_inverted_index(column_name, _inverted_index_iterators[cid], &output_result);
+            if (res.ok()) {
+                erased_preds.emplace_back(pred);
+            }
+        }
+    }
+
+    // ---------------------------------------------------------
+    // Erase predicates that hit bitmap index.
```
Comment (srlch) [inline be/src/storage/rowset/segment_iterator.cpp:1887]: hit inverted index?

File: be/src/storage/rowset/segment_iterator.cpp [hunk 2]
```diff
@@ -1834,6 +1844,57 @@ Status SegmentIterator::_apply_del_vector() {
...
+        roaring::Roaring output_result = row_bitmap;
+        InvertedIndexIterator* inverted_iter = _inverted_index_iterators[cid];
+        if (inverted_iter == nullptr) {
+            continue;
+        }
+        const FieldPtr& fieldPtr = _schema.field(cid);
+        std::string_view sv_c = fieldPtr->name();
+        std::string column_name = {sv_c.begin(), sv_c.end()};
+        for (const ColumnPredicate* pred : pred_list) {
+            Status res = pred->seek_inverted_index(column_name, _inverted_index_iterators[cid], &output_result);
```
Comment (srlch) [inline be/src/storage/rowset/segment_iterator.cpp:1879]: It seems that output_result is just a local variable. Should we use the output_result to update the _scan_range?

File: be/src/storage/rowset/segment_iterator.cpp [hunk 3]
```diff
@@ -1834,6 +1844,57 @@ Status SegmentIterator::_apply_del_vector() {
...
+
+    // ---------------------------------------------------------
+    // Erase predicates that hit bitmap index.
+    // ---------------------------------------------------------
+    for (const ColumnPredicate* pred : erased_preds) {
+        PredicateList& pred_list = _opts.predicates[pred->column_id()];
+        pred_list.erase(std::find(pred_list.begin(), pred_list.end(), pred));
+    }
+
+    _opts.stats->rows_inverted_index_filtered += (input_rows - row_bitmap.cardinality());
```
Comment (srlch) [inline be/src/storage/rowset/segment_iterator.cpp:1894]: It seems that it always be 0 for input_rows - row_bitmap.cardinality()?

File: fe/fe-core/src/main/java/com/starrocks/analysis/MatchPredicate.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+ * Fulltext match predicate
+ */
+public class MatchPredicate extends Predicate {
+    private static final Logger LOG = LogManager.getLogger(MatchPredicate.class);
+
+    public enum Operator {
+        MATCH_TERMS("MATCH_TERMS", "match_terms", TExprOpcode.MATCH_TERMS),
+        MATCH_ALL("MATCH_ALL", "match_all", TExprOpcode.MATCH_ALL),
+        MATCH_PHRASE("MATCH_PHRASE", "match_phrase", TExprOpcode.MATCH_PHRASE),
+        MATCH_FUZZY("MATCH_FUZZY", "match_fuzzy", TExprOpcode.MATCH_FUZZY),
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MatchPredicate.java:47]: 1. Is the `MATCH_FUZZY` same with `like`, `MATCH_WILDCARD` same with `regex_like`?
2. I send a query `select * from test where match_fuzzy(f1, "thinkbad", 1)`. If there is no inverted index of f1, what's the behavior?
3. When to determine use inverted index or not if `f1` already has an inverted index? Depends on the cardinality of `f1` or use the inverted index in any cardinality?
4. `select * from test where match_fuzzy(f1, "thinkbad", 1) and f2 = 'xxx'` If `f1` has inverted index, `f2` has bitmap index, how to determine the filter order?

File: be/CMakeLists.txt [hunk 5]
```diff
@@ -174,6 +174,17 @@ if (${JEMALLOC_DEBUG})
 endif()
 
 # Set all libraries
+add_library(ic STATIC IMPORTED)
```
Comment (chaoyli) [inline be/CMakeLists.txt:177]: It's better to describe the solution and effect in the commit message.
1. The solution of inverted index.
2. Under inverted index, how much performance gained, how much storage cost.

**Comment by kevincai** [comment]
@dujijun007 suppose this PR is replaced by https://github.com/StarRocks/starrocks/pull/49346, we can close this PR, right?

**Pull request closed without merging**